### PR TITLE
Issue_#1236(better solution) : "npm run serve" now logs warning and usefull info and also show compilation progress

### LINKED ISSUE #1236: The `npm run serve` command hides almost all logging
**Describe the bug**
The `npm run serve` command hides almost all logging

**To Reproduce**
Run the `npm run serve` command

**Expected behavior**
This behavior is not acceptable.

1. There is no feedback as to what is happening. 
    1. There is no useful troubleshooting information shown
    1. There is no indication of the state of progress

Neither the development nor DevOps teams will therefore have the best information to troubleshoot issues. We require:
1. Full logging to the screen indicating what is happening
1. No clearing of the output between messages

**Actual behavior**
1. First this message is seen
image
1. Then the message is cleared, then this message is seen for a few seconds, then disappears
image
1. The screen then goes blank for several minutes with no indication of what is happening. The user may feel that the operation is failing.
1. This message is then displayed for several minutes with no indication of any activity happening
image
1. Finally, the screen clears once more then this message is visible
image

**Additional details**
- N/A

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by Tarunmeena0901**
**What kind of change does this PR introduce?**
bugfix , feature

**Issue Number:**

Fixes #1236 

**Did you add tests for your changes?**
NO

**Snapshots/Videos:**
Screenshot 2023-12-24 161700

Screenshot 2023-12-24 161753

**If relevant, did you update the documentation?**
YES i updated the installation.md file

**Summary**
so , whenever a developer was running npm run serve , there were no usefull logs during compile time or run time also there was no indicator for progress which can tell how much project is compiled .
there was no usefull troubleshooting information shown for the development or DevOps teams

**Does this PR introduce a breaking change?**
The integration of a state progress indicator in the webpack configuration and console output enhances the development experience by providing real-time feedback on compilation progress and runtime errors.

**Other information**

1. integrated an custom webpack progress plugin in webpack.config.js

2. changed the infrastructureLogging in webpack.config

3. **the bigger part of this issue :**
- as our project is bootstraped using `create-react-app` there is very less custom configuration change one can do . so the only
way is to make changes inside `node_modules` , in our case it was "node_modules/react-scripts/config/webpack.config.js".
but as we know even if u change files inside node_module you cannot commit them so u have to make your own custom
configuration file but making our whole webpack configuration is lot of work so one solution i came up with is to make a
separate `config-override` file in root directory and use `react-app-rewired` to change the webpack configuration during
compile time itself.
inside `config-override` folder u can see a `package.json` which is made so the commonJs module of react-app-rewired library
dont conflict with ES module of our project

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

YES

### CODE REVIEW

**Comment by Tarunmeena0901** [comment]
Now for the two points you mentioned in my previous pull request.

1. Is this a viable long term solution ?
- not everybody needs extra logs that's why added options in  .env file to if a developer need logs or not just like in our talawa-api logs . 
- also updated the documents for ease of understanding
- if a person choose "YES" for logs , react-app-rewired will be used to just to overwrite the `webpack.config.js` but under the hood our app will run same as before . this is much safer option for this solution as we are not ejecting our application and we will still get benefits from `Create-react-app`

2. The API users typescript and doesn't have this issue. Have you looked there to see what other options could be used?
- for talawa-api , it uses a logger package name `winston` which allow you log info , warning and error at you will but it wont help us in case of talawa admin as it is bootstrapped from `create-react-app` which dont allow us to change and customize configuration without ejecting our application (ejecting is a bad idea )
-  under the hood webpack run as a bundler so our compilation and runtime logs are provided by webpack so we have to make changes in `webpack.config` for that we need to overwrite the `webpack.config` using `react-app-rewired` (as the create-react-app dont allow modification), whereas our talawa-api backend dont use webpack and so we have all controls of logs . 
- thats why talawa-api solution wont work in this case. also this issue have nothing to do with typescript.

your questions helped me understand things much more deeper please ask more if you have any 😄

**Comment by palisadoes** [comment]
@noman2002 @DMills27 Your input would be appreciated here. We are trying to get better logging and need second opinions on the proposed approach here

File: .env.example [hunk 1]
```diff
@@ -20,4 +20,8 @@ REACT_APP_USE_RECAPTCHA=
 REACT_APP_RECAPTCHA_SITE_KEY=
 
 # has to be inserted in the env file to use plugins and other websocket based features.
-REACT_APP_BACKEND_WEBSOCKET_URL=ws://localhost:4000/graphql
\ No newline at end of file
+REACT_APP_BACKEND_WEBSOCKET_URL=ws://localhost:4000/graphql
+
+# If you want to logs Compiletime and Runtime error , warning and info write YES or if u want to  
+# keep the console clean leave it blank
+LOGS_CONFIG_OVERRIDE=""
```
Comment (noman2002) [inline .env.example:27]: "LOGS_CONFIG_OVERRIDE" is not self explanatory. please make it something easy to understand. Something like "SHOW_LOGS", "ALLOW_LOGS", etc.

**Review by noman2002 — CHANGES_REQUESTED** [review]

File: .env.example (hunk 1 above)
Comment (Tarunmeena0901) [reply .env.example:27]: yes sir I also think that would be better and less confusing.

**Review by noman2002 — APPROVED** [review]
Looks fine to me.

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. Please move the `config-overrides` directory to be a sub-directory of `scripts`
2. I think the existing scripts are related to GitHooks so move them to a `scripts/githooks` directory and update the hooks.

We have done this with the API, and the consistency will be good.

**Comment by Tarunmeena0901** [comment]
> 1. Please move the `config-overrides` directory to be a sub-directory of `scripts`
> 2. I think the existing scripts are related to GitHooks so move them to a `scripts/githooks` directory and update the hooks.
> 
> We have done this with the API, and the consistency will be good


- Yes sir, I agree that keeping the consistency through out all the repository is a good practice and also help other developers in understanding the codebase better . **BUT** sir the problem here with the dependency we are using i.e react-app-rewired, is that `React-app-rewired` actually uses standard `node.js` name resolution, in that it imports `path/to/config-overrides` and not `path/to/config-overrides.j`s , This means that it needs to be a directory named config-overrides containing an `index.js` inside it, so we can not move the config-override inside scripts directory as it will not work. 

- one thing we can do is i can make another sub-directory inside `./config-override`s named as `./config-overrides/scripts` and transfer the scripts from `./config-overrides` to `./config-overrides/scripts` 
   or else
 we can keep it as it is

**Comment by palisadoes** [comment]
Why can't the path be `scripts/githooks/config -overrides#`? Putting scripts in the root directory is not good practice

**Comment by Tarunmeena0901** [comment]
> Why can't the path be `scripts/githooks/config -overrides#`? Putting scripts in the root directory is not good practice

Its done !!! although i was not able find the solution before but looking into the node_modules file of dependency helped me change the default path on my own, 
but I have one suggestion sir @palisadoes that these scripts have nothing to do with `git hooks` so it would be better if we keep this in `./scripts/config-overrides` rather than `scripts/githooks/config-overrides` .  please see recent commit and provide your view on this

**Comment by palisadoes** [comment]
> > Why can't the path be `scripts/githooks/config -overrides#`? Putting scripts in the root directory is not good practice
> 
> Its done !!! although i was not able find the solution before but looking into the node_modules file of dependency helped me change the default path on my own, but I have one suggestion sir @palisadoes that these scripts have nothing to do with `git hooks` so it would be better if we keep this in `./scripts/config-overrides` rather than `scripts/githooks/config-overrides` . please see recent commit and provide your view on this

OK

**Comment by palisadoes** [comment]
@noman2002 @DMills27 Please review

**Review by noman2002 — APPROVED** [review]
Looks fine to me.  @DMills27 Please have a look.

**Review by DMills27 — DISMISSED** [review]

**Comment by palisadoes** [comment]
Please fix the failing linting tests then we can approve and merge

**Comment by Tarunmeena0901** [comment]
> Please fix the failing linting tests then we can approve and merge

@palisadoes failing linting checks FIXED 💯 💯 💯

**Pull request merged**